File: synth/synth/mutators/wrong_math.py
"""Semantic math error mutators — ground truth records displayed values."""

from __future__ import annotations
import random


def apply_wrong_iva_sum(spec, gt, rng: random.Random) -> tuple:
    """Offset the displayed grand total by a small random amount."""
    error = round(rng.uniform(0.02, 8.00), 2)
    direction = rng.choice([-1, 1])
    spec.totals.displayed_grand_total = round(
        spec.totals.displayed_grand_total + direction * error, 2
    )
    # Ground truth stores what's DISPLAYED — extractor will see the wrong number
    gt.total_amount = spec.totals.displayed_grand_total
    gt._meta.mutations.append("wrong_iva_math")
    return spec, gt


def apply_wrong_line_total(spec, gt, rng: random.Random) -> tuple:
    """Make one line item's displayed total not match qty × unit_price."""
    if not spec.items:
        return spec, gt
    idx = rng.randint(0, len(spec.items) - 1)
    error = round(rng.uniform(0.01, 2.00), 2)
    spec.items[idx]["total_price"] = round(
        spec.items[idx]["total_price"] + rng.choice([-1, 1]) * error, 2
    )
    # Ground truth reflects displayed value
    gt.line_items[idx].total_price = spec.items[idx]["total_price"]
    gt._meta.mutations.append("line_total_wrong")
    return spec, gt


MUTATOR_MAP = {
    "wrong_iva_math":    apply_wrong_iva_sum,
    "line_total_wrong":  apply_wrong_line_total,
}

File: synth/js/mutators/wrong_math.mjs
export function applyWrongIvaSum(spec, gt, rng) {
  if (!spec.totals) return; // albaran has no totals
  const error = Math.round(rng.uniform(0.02, 8.00) * 100) / 100;
  const sign = rng.choice([-1, 1]);
  spec.totals.displayed_grand_total = Math.round(
    (spec.totals.displayed_grand_total + sign * error) * 100
  ) / 100;
  gt.total_amount = spec.totals.displayed_grand_total;
  gt._meta.mutations.push('wrong_iva_math');
}

export function applyWrongLineTotal(spec, gt, rng) {
  if (!spec.items.length) return;
  const idx = rng.randint(0, spec.items.length - 1);
  const error = Math.round(rng.uniform(0.01, 2.00) * 100) / 100;
  const sign = rng.choice([-1, 1]);
  spec.items[idx].total_price = Math.round((spec.items[idx].total_price + sign * error) * 100) / 100;
  gt.line_items[idx].total_price = spec.items[idx].total_price;
  gt._meta.mutations.push('line_total_wrong');
}

export const MUTATOR_MAP = {
  wrong_iva_math:   applyWrongIvaSum,
  line_total_wrong: applyWrongLineTotal,
};
